modules: openthread: add gpio diag command implementation

maciejbaczmanski opened the pull request:
This PR implements `ot diag gpio get`, `ot diag gpio set` and `ot diag gpio mode` commands.
It also adds additional `gpio_pin_is_input_dt` and `gpio_pin_is_output_dt` helpers.

To enable gpio diag commands, in Devicetree create `openthread` node in `/options/` path with `compatible = "openthread,config"` property and `diag-gpios` property, which should contain array of GPIO pin's configuration properties containing controller phandles, pin numbers and pin flags. e.g:
 
```
  options {
 		openthread {
 			compatible = "openthread,config";
 			diag-gpios = <&gpio0 0 GPIO_ACTIVE_HIGH>,
 				     <&gpio1 0 GPIO_ACTIVE_LOW>;
 		};
 	};
 ```
To enable reading current gpio pin mode, define `CONFIG_GPIO_GET_DIRECTION` in prj.conf.

Note: `<gpio>` in `diag gpio` commands is an index of diag-gpios array. For example shown above, `ot diag gpio mode 0` will return current mode of pin nmb 0 controlled by `gpio0` controller.

Signed-off-by: Maciej Baczmanski <maciej.baczmanski@nordicsemi.no>

maciejbaczmanski committed:
net: openthread: add gpio diag command implementation

implemented ot diag gpio get, set and mode commands

Signed-off-by: Maciej Baczmanski <maciej.baczmanski@nordicsemi.no>

rlubos commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +73,151 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+static const struct device *const gpio_port_dev[] = {
+	DT_GPIO_PORT_FROM_PROP_IDX(DT_PATH(zephyr_user), openthread_gpio_ports, 0)};
+#else
+static const struct device *const gpio_port_dev[] = {DT_FOREACH_PROP_ELEM_SEP(
```
Can you elaborate on why wouldn't this approach work for `DT_PROP_LEN(DT_PATH(zephyr_user), openthread_gpio_ports) == 1`?

maciejbaczmanski replied:
According to documentation of ` DT_FOREACH_PROP_ELEM_SEP(node_id,prop,fn,sep)` macro: 
`The "prop" argument must refer to a property with type string, array, uint8-array, string-array, phandles, or phandle-array. It is an error to use this macro with properties of other types.`
If there is only one port in devicetree e.g.:
```
zephyr,user {
		openthread-gpio-ports = <&gpio0>;
	};
```
`prop` is type of phandle (singular), hence `DT_FOREACH_PROP_ELEM_SEP` fails at build

pdgendt replied:
It would be nice if we could simply list all devices with a `gpio-controller;` property from `DT_`-macros.

edmont replied:
Would it be possible to require `openthread-gpio-ports` to be declared as an array?

maciejbaczmanski replied:
I have discovered that the issue was laying in creating a property in `zephyr,user` node. It's type was changed automatically on build, so if there was only one element, it changed type to `phandle` (singular). Creating a separate node with Devicetree binding has solved this issue.

rlubos commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +73,151 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+	DT_PATH(zephyr_user), openthread_gpio_ports, DT_GPIO_PORT_FROM_PROP_IDX, (, ))};
+#endif /*DT_PROP_LEN(DT_PATH(zephyr_user), openthread_gpio_ports) == 1*/
+
+const struct device *const otGpioGetDevAndPin(uint32_t aGpio, gpio_pin_t *pin, otError *error)
```
This should be static? Also, for non-OT defined APIs, please use Zephyr coding style (i. e. snake_case).

rlubos commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +73,151 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+	const struct device * dev = NULL;
+	uint32_t port;
+
+	port = aGpio >> 5;
+	*pin = aGpio & 0x1F;
```
Could define symbols for those magic numbers.

pdgendt replied:
```suggestion
	port = aGpio / 32;
	*pin = aGpio % 32;
```

pdgendt commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +75,159 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+	DT_PATH(zephyr_user), openthread_gpio_ports, DT_GPIO_PORT_FROM_PROP_IDX, (,))};
+#endif /*DT_PROP_LEN(DT_PATH(zephyr_user), openthread_gpio_ports) == 1*/
+
+static const struct device *const gpio_get_dev_and_pin(uint32_t aGpio,
+							gpio_pin_t *pin,
+							otError *error)
```
Nit: fix indentation

edmont commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +75,158 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+
+/*
+ * To enable gpio diag commands, create `openthread-gpio-ports` property
+ * in `zephyr,user` node.
```
```suggestion
 * in Device Tree's `zephyr,user` node.
```

edmont commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +75,158 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+#else
+static const struct device *const gpio_port_dev[] = {DT_FOREACH_PROP_ELEM_SEP(
+	DT_PATH(zephyr_user), openthread_gpio_ports, DT_GPIO_PORT_FROM_PROP_IDX, (,))};
+#endif /*DT_PROP_LEN(DT_PATH(zephyr_user), openthread_gpio_ports) == 1*/
```
```suggestion
#endif /* DT_PROP_LEN(DT_PATH(zephyr_user), openthread_gpio_ports) == 1 */
```

edmont commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +75,158 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+	DT_PATH(zephyr_user), openthread_gpio_ports, DT_GPIO_PORT_FROM_PROP_IDX, (,))};
+#endif /*DT_PROP_LEN(DT_PATH(zephyr_user), openthread_gpio_ports) == 1*/
+
+static const struct device *const gpio_get_dev_and_pin(uint32_t aGpio, gpio_pin_t *pin,
```
For me it would seem more intuitive to return the error and pass the device as a parameter to this function. Thoughts?

edmont reviewed (commented):
Thanks, a few comments.

edmont reviewed (dismissed):
Great, thanks!

pdgendt commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +75,175 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
 }
+
+/*
+ * To enable gpio diag commands, in Devicetree create `gpio_ports` node in root path
+ * with `gpio-ports-phandles` property, which should contain phandles of GPIO port nodes.
+ * It is important to create correct Devicetree binding (more about bindings location:
+ * https://docs.zephyrproject.org/latest/build/dts/bindings.html#where-bindings-are-located),
+ * which defines `gpio-ports-phandles`
+ * type as phandles. e.g:
+ *
+ * Devicetree:
+ *
+ * gpio-ports {
+ *		compatible = "gpio-ports";
+ *		gpio-ports-phandles = <&gpio0 &gpio1>;
+ * };
+ *
+ * Devicetree binding:
+ *
+ * description: Node which contains GPIO port phandles
+ *
+ * compatible: "gpio-ports"
+ *
+ * properties:
+ *   gpio-ports-phandles:
+ *     type: phandles
+ *     required: true
```
This does not exist AFAIK? It shouldn't be up to users to add the binding, right?

maciejbaczmanski replied:
I've created a binding yaml file and updated description.

carlescufi commented:
@cfriedt @gmarull @henrikbrixandersen would you mind reviewing the new GPIO binding?

gmarull commented on modules/openthread/platform/diag.c:
```diff
@@ -5,9 +5,16 @@
 #include "platform-zephyr.h"
+#include "utils/code_utils.h"
+
+#define GPIO_PINS_PER_PORT	32
```
there's `GPIO_MAX_PINS_PER_PORT` defined in `drivers/gpio.h`

gmarull commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +75,159 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+ * gpio = (port * 32) + pin
+ */
+#if DT_NODE_EXISTS(DT_PATH(gpio_ports)) && \
+	DT_NODE_HAS_PROP(DT_PATH(gpio_ports), gpio_ports_phandles)
```
last check is redundant, since DT prop is flagged as required, so build will not even start if it's missing.

gmarull commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +75,159 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+ * with `compatible = "gpio-ports"` property and `gpio-ports-phandles` property,
+ * which should contain phandles of GPIO port nodes. e.g:
+ *
+ * gpio-ports {
+ *		compatible = "gpio-ports";
```
If this is something specific to OT, this compatible name is too generic. And this is actually not a hardware description, but a software choice for openthread. @mbolivar-nordic advice here? Can `chosen` node props take phandles?

maciejbaczmanski replied:
The point of a generic name was that this binding can be used in other places. For example for drivers/gpio/gpio_shell.c to replace/extend  current implementation which uses device_get_binding() (which requires knowledge of node's full name) to get device struct.

maciejbaczmanski replied:
@gmarull I've done some additional research about moving phandles to `chosen` node. I think that it is not possible. According to documentation: https://docs.zephyrproject.org/apidoc/latest/group__devicetree-generic-chosen.html `DT_CHOSEN` macro returns a "node identifier for the chosen node property" (singular). I tried to put the phandles into a `chosen` property in overlay file anyways and got a warning: `Properties in chosen must be references to nodes`. Trying to use `DT_FOREACH_PROP_ELEM_SEP` macro on `chosen` node results in an error at build time.

mbolivar-nordic replied:
> @mbolivar-nordic advice here? Can chosen node props take phandles?

As discovered already, no.

Otherwise it's a bindings question and I'm not involved in reviewing bindings at the moment; ask @galak

gmarull commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +74,156 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+	return error;
+}
+
+otError otPlatDiagGpioSetMode(uint32_t aGpio, otGpioMode aMode)
```
It looks like we just need a diagnostic GPIO, so I think something like this https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/options/u-boot.yaml would make more sense:

```
options {
    openthread {
        compatible = "openthread,config";
        diag-gpios = <&gpio0 1 ...>;
    };
};

gmarull replied:
@sjg20 could you comment on that since you're the maintainer of https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/options.yaml ?

maciejbaczmanski replied:
The issue is that `/options` node has not been used in Zephyr before and doesn't a proper description yet. Do you think we could require `openthread` node to have a label, e.g:

``` 
openthread_config: openthread {
        compatible = "openthread,config";
        diag-gpios = <&gpio0 3  GPIO_ACTIVE_HIGH>,
                     <&gpio1 11 GPIO_ACTIVE_HIGH>;
};
```

then we could use `DT_NODELABEL(openthread_config)` in diag.c and leave node's placement up to user for now. Later we could discuss proper placement for this type of nodes as a separate task?

gmarull replied:
Even if not defined, it can be created. Whether additional DT helpers for such node should exist needs to be decided.

maciejbaczmanski committed:
gpio: add `gpio_pin_is_input_dt` and `gpio_pin_is_output_dt` helpers

Add helpers to allow providing a `struct gpio_dt_spec` directly
to `gpio_pin_is_input` and `gpio_pin_is_output` functions.

Signed-off-by: Maciej Baczmanski <maciej.baczmanski@nordicsemi.no>

pdgendt commented on dts/bindings/options/openthread,config.yaml:
```diff
@@ -0,0 +1,24 @@
+    description: |
+      This enables access to diagnostic GPIO pins. Each field consists of
+      GPIO pin's configuration: controller's phandle, pin number and configuration flags.
+    required: true
```
Does this have to be `required`? I imagine we could add properties in the future, which could all be optional.

pdgendt commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +71,153 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+ * Note: `<gpio>` in `diag gpio` commands is an index of diag-gpios array. For example shown above,
+ * `ot diag gpio mode 0` will return current mode of pin nmb 0 controlled by `gpio0` controller.
+ */
+#if DT_NODE_EXISTS(DT_PATH(options, openthread))
```
I guess this would be the default path to put the properties, but I think the `compatible` would be preferred to match the node?

pdgendt commented on dts/bindings/options/openthread,config.yaml:
```diff
@@ -0,0 +1,24 @@
+    description: |
+      This enables access to diagnostic GPIO pins. Each field consists of
+      GPIO pin's configuration: controller's phandle, pin number and configuration flags.
+    required: false
```
Nit: `required: false` can be omitted.

pdgendt commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +71,157 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+	DT_FOREACH_PROP_ELEM_SEP(DT_COMPAT_GET_ANY_STATUS_OKAY(openthread_config),
+				 diag_gpios, GPIO_DT_SPEC_GET_BY_IDX, (,))};
+
+static otError gpio_get_spec(uint32_t aGpio, const struct gpio_dt_spec **spec)
```
Nit: zephyr functions use `snake_case` for arguments.

pdgendt commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +71,157 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+otError otPlatDiagGpioSet(uint32_t aGpio, bool aValue)
+{
+	const struct gpio_dt_spec *spec;
+	otError error = OT_ERROR_NONE;
```
No need to initialise `error`.

maciejbaczmanski commented:
@gmarull would you mind reviewing the changes?

maciejbaczmanski commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +71,157 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+	*spec = &gpio_spec[gpio_idx];
+
+	otEXPECT_ACTION(otPlatDiagModeGet(), error = OT_ERROR_INVALID_STATE);
+	otEXPECT_ACTION(device_is_ready((*spec)->port), error = OT_ERROR_INVALID_ARGS);
```
after `gpio_is_ready_dt()` helper implementation this line can be simplified

maciejbaczmanski commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +71,157 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+
+#if defined(CONFIG_GPIO_GET_DIRECTION)
+	rv = gpio_pin_is_input_dt(spec);
+	otEXPECT_ACTION(rv == 1, error = OT_ERROR_FAILED);
```
```suggestion
	otEXPECT_ACTION(rv == 1, error =OT_ERROR_INVALID_STATE);
```
to comply with https://github.com/openthread/openthread/pull/8538

maciejbaczmanski commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +71,157 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+
+#if defined(CONFIG_GPIO_GET_DIRECTION)
+	rv = gpio_pin_is_output_dt(spec);
+	otEXPECT_ACTION(rv == 1, error = OT_ERROR_FAILED);
```
```suggestion
	otEXPECT_ACTION(rv == 1, error = OT_ERROR_INVALID_STATE);
```
to comply with https://github.com/openthread/openthread/pull/8538

pdgendt commented on modules/openthread/platform/diag.c:
```diff
@@ -68,3 +71,157 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
+ * which should contain array of GPIO pin's configuration properties containing controller phandles,
+ * pin numbers and pin flags. e.g:
+ *
+ * options {
```
Nit: fix indentation
